### ISSUE TITLE
Asynchronous programming. Python3.5

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#4622]
Asynchronous programming. Python3.5+<br>
<br>
https://ift.tt/2C1K4Hv<br>
<br>
<br>
<br>
<div>
<div>
<img alt="Asynchronous programming. Python3.5+" src="https://luminousmen.com/media/asynchronous-programming-python3.5.jpg">
<p>This is a practical post of the series of asynchronous programming.</p>
<p>Whole series:</p>
<p>In this post we will be talking about the Python stack on the concepts we talked so far: from the simplest like threads, processes to the asyncio library.</p>
<p>Asynchronous programming in Python has become more and more popular lately. There are many different libraries in python for doing asynchronous programming. One of those libraries is asyncio, which is a python standard library added in Python 3.4. In Python 3.5 we got an async/await syntax. Asyncio is part of the reason asynchronous programming is becoming more popular in Python. This article will explain what asynchronous programming is and compare some of these libraries.</p>
<hr>
<h2>Quick Recap</h2>
<p>What we have realized so far from the previous posts:</p>
<ul>
<li>
<p><strong>Sync:</strong> Blocking operations.</p>
</li>
<li>
<p><strong>Async:</strong> Non-blocking operations.</p>
</li>
<li>
<p><strong>Concurrency:</strong> Making progress together.</p>
</li>
<li>
<p><strong>Parallelism:</strong> Making progress in parallel.</p>
</li>
<li>
<p><strong>Parallelism implies Concurrency</strong>. But <a href="https://luminousmen.com/post/concurrency-and-parallelism-are-different">Concurrency doesn&rsquo;t always mean Parallelism</a>.</p>
</li>
</ul>
<hr>
<p>Python code can now be mainly executed in one of two worlds, synchronous or asynchronous. You should think of them as <strong>separate worlds with different libraries and styles of calls</strong>, but sharing variables and syntax.</p>
<p>In the synchronous Python world, which has existed for several decades, you call functions directly, and everything is processed sequentially, exactly as you wrote your code. There are options to run the code concurrently.</p>
<h2>Synchronous world</h2>
<p>In this post we will be comparing different implementations of the same code. We will try to execute two functions. First one is calculating the power of number:</p>
<pre>
<code>def cpu_bound(a, b):
 return a ** b
 </code>
</pre>
<p>We will do it N times:</p>
<pre>
<code>def simple_1(N, a, b):
 for i in range(N):
 cpu_bound(a, b)
 </code>
</pre>
<p>And the second one is downloading data from the web:</p>
<pre>
<code>def io_bound(urls):
 data = []
 for url in urls:
 data.append(urlopen(url).read())
 return data
 
 def simple_2(N, urls):
 for i in range(N):
 io_bound(urls)
 </code>
</pre>
<p>To compare how much time function works we implement simple decorator/context manager:</p>
<pre>
<code>import time
 from contextlib import ContextDecorator
 
 class timeit(object):
 def __call__(self, f):
 @functools.wraps(f)
 def decorated(*args, **kwds):
 with self:
 return f(*args, **kwds)
 return decorated
 
 def __enter__(self):
 self.start_time = time.time()
 
 def __exit__(self, *args, **kw):
 elapsed = time.time() - self.start_time
 print("{:.3} sec".format(elapsed))
 </code>
</pre>
<p>Now let's put it all together and run, to understand how much time my machine will be executing this code:</p>
<pre>
<code>import time
 import functools
 from urllib.request import urlopen
 from contextlib import ContextDecorator
 
 
 class timeit(object):
 def __call__(self, f):
 @functools.wraps(f)
 def decorated(*args, **kwds):
 with self:
 return f(*args, **kwds)
 return decorated
 
 def __enter__(self):
 self.start_time = time.time()
 
 def __exit__(self, *args, **kw):
 elapsed = time.time() - self.start_time
 print("{:.3} sec".format(elapsed))
 
 
 def cpu_bound(a, b):
 return a ** b
 
 
 def io_bound(urls):
 data = []
 for url in urls:
 data.append(urlopen(url).read())
 return data
 
 
 @timeit()
 def simple_1(N, a, b):
 for i in range(N):
 cpu_bound(a, b)
 
 
 @timeit()
 def simple_2(N, urls):
 for i in range(N):
 io_bound(urls)
 
 
 if __name__ == '__main__':
 a = 7777
 b = 200000
 urls = [
 "http://google.com",
 "http://yahoo.com",
 "http://linkedin.com",
 "http://facebook.com"
 ]
 simple_1(10, a, b)
 simple_2(10, urls)
 </code>
</pre>
<p>We implemented execution of the same functions <code>N</code> times sequentially.</p>
<p>On my hardware, <code>cpu_bound</code> function took 2.18 sec, <code>io_bound</code> &mdash; 31.4 sec.</p>
<p>So, we get our basic performance. Let's move on to threads.</p>
<h3>Threads</h3>
<p><img alt="threads" src="https://luminousmen.com/media/asynchronous-programming-python3.5-1.jpg"></p>
<p>A thread is the smallest unit of processing that can be performed in an OS.</p>
<p>Threads of a process can <strong>share the memory of global variables</strong>. If a global variable is changed in one thread, this change is valid for all threads.</p>
<p>In simple words, a <strong>thread</strong> is a sequence of such operations within a program that can be executed independently of other code.</p>
<p>Threads executing <strong>concurrently</strong> but can be executing in parallel &mdash; it depends on the system on which they are running.</p>
<p>Python threads are implemented using OS threads in all implementations I know (CPython, PyPy and Jython). For each Python thread, there is an underlying OS thread.</p>
<p>One thread is executed on one processor core per unit of time. It works until it consumes its time slice (default is 100 ms) or until it gives up control to the next thread by making a system call.</p>
<p>Let's try to implement our example functionality using threads:</p>
<pre>
<code>from threading import Thread
 
 @timeit()
 def threaded(n_threads, func, *args):
 jobs = []
 for i in range(n_threads):
 thread = Thread(target=func, args=args)
 jobs.append(thread) 
 
 # start the threads
 for j in jobs:
 j.start() 
 
 # ensure all of the threads have finished
 for j in jobs:
 j.join()
 
 if __name__ == '__main__':
 ...
 threaded(10, cpu_bound, a, b)
 threaded(10, io_bound, urls)
 </code>
</pre>
<p>On my hardware, <code>cpu_bound</code> took 2.47 sec, <code>io_bound</code> &mdash; 7.9 sec.</p>
<p>The I/O-bound function executed more than 5 times faster because we download the data in parallel on separate threads. But why CPU-bound operation goes slower?</p>
<p>In the reference implementation of Python &mdash; CPython there is the infamous GIL (Global Interpreter Lock). And we slowly go to its section...</p>
<h3>Global Interpreter Lock (GIL)</h3>
<p>First of all, GIL is a lock that must be taken before any access to Python (and this is not only the execution of Python code but also calls to the Python C API). In essence, <strong>GIL is a global semaphore</strong> that does not allow more than one thread to work simultaneously within an interpreter.</p>
<p>Strictly speaking, the only calls available after running the interpreter with an uncaptured GIL are its capture. Violation of the rule leads to an instant crash (the best option) or delayed crash of the program (much worse and harder to debug).</p>
<p><em>How it works?</em></p>
<p>When the thread starts, it performs a GIL capture. After a while, the process scheduler decides that the current thread has done enough and give control to the next thread. Thread #2 sees that GIL is captured so that it does not continue to work, but plunges itself into sleep, yielding the processor to thread #1.</p>
<p>But the <strong>thread cannot hold GIL indefinitely</strong>. Prior to Python 3.3, GIL switched every 100 machine code instructions. In later versions of GIL, a thread can be held no longer than 5 ms. GIL is also released if the thread makes a system call, works with a disk or network(I/O-bound operation).</p>
<p>In fact, GIL in python makes the idea of &#8203;&#8203;using threads for parallelism in computational problems(CPU-bound operations) useless. They will work sequentially even on a multiprocessor system. <strong>On CPU-bound tasks, the program will not accelerate</strong>, but only slow down, because now the threads will have to halve the processor time. At the same time, the <strong>GIL I/O operation will not slow down</strong>, since before the system call the thread releases the GIL.</p>
<p>It is clear that GIL slows down the execution of our program due to the additional work of creating and communicating threads, capturing and releasing the semaphore itself and preserving the context. But it needs to be mentioned that GIL <strong>does not limit</strong> parallel execution.</p>
<p>GIL is not part of the language and does not exist in all language implementations, but only in the above mentioned CPython.</p>
<p><em>So why the heck does he even exist?</em></p>
<p><strong>GIL protects operating data structures from concurrent access problems</strong>. For example, it prevents the race condition when the object's reference count value changes. GIL makes it easy to integrate non-thread safe libraries on C. Thanks to GIL, we have so many fast modules and binders for almost everything.</p>
<p>There are exceptions, for example, the GIL control mechanism is available to C libraries. For example, <code>NumPy</code> releases it on long operations. Or, when using the <code>numba</code> package, the programmer can control the semaphore disabling itself.</p>
<p>On this sad note, you can come to the conclusion that threads will be enough for parallelizing tasks that are tied to I/O. But computing tasks should be run in separate processes.</p>
<h3>Processes</h3>
<p>From the OS point of view, a process is a data structure that holds a memory area and some other resources, for example, files opened by it. Often the process has one thread, called <strong>main</strong>, but the program can create any number of threads. At the start, the thread is not allocated individual resources, instead, it uses the memory and resources of the process that spawned it. Due to this, the threads can quickly start and stop.</p>
<p>Multitasking is handled by the scheduler &mdash; part of the OS kernel, which in turn loads the execution threads into the central processor.</p>
<p>Like threads, processes are always executed <strong>concurrently</strong>, but they can also be executed in parallel, depending on the presence of the hardware component.</p>
<pre>
<code>from multiprocessing import Process
 
 @timeit()
 def multiprocessed(n_threads, func, *args):
 processes = []
 for i in range(n_threads):
 p = Process(target=func, args=args)
 processes.append(p)
 
 # start the processes
 for p in processes:
 p.start()
 
 # ensure all processes have finished execution
 for p in processes:
 p.join()
 
 if __name__ == '__main__':
 ...
 multiprocessed(10, cpu_bound, a, b)
 multiprocessed(10, io_bound, urls)
 </code>
</pre>
<p>On my hardware, <code>cpu_bound</code> took 1.12 sec, <code>io_bound</code> &mdash; 7.22 sec.</p>
<p>So, the calculation operation executed faster than threaded implementation because now we are not stuck in capturing GIL, but the I/O bound function took slightly more time because processes are more heavyweight than threads.</p>
<h2>Asynchronous world</h2>
<p><img alt="asynchronous programming" src="https://luminousmen.com/media/asynchronous-programming.jpg"></p>
<p>In an asynchronous world, everything changes a bit. Everything works in a central event-processing loop, which is a small code that allows you to run several coroutines (an important term, in simple terms it is not OS-managed threads, except they are co-operatively multitasking, and hence not truly concurrent) at the same time. Coroutines work synchronously until the expected result is reached, and then they stop, transfer control to the event loop, and something else can happen.</p>
<h3>Green threads</h3>
<p><img alt="Green threads" src="https://luminousmen.com/media/asynchronous-programming-python3.5-2.jpg"></p>
<p><a href="https://luminousmen.com/post/asynchronous-programming-await-the-future">Green threads</a> are a primitive level of asynchronous programming. A green thread is a regular thread, except that switching between threads is done in the application code(on the user level), and not in the processor(OS level). The core of it is <a href="https://luminousmen.com/post/asynchronous-programming-blocking-and-non-blocking">non-blocking operations</a>. Switching between threads occurs only on I/O. Non-I/O threads will take control forever.</p>
<p><a href="http://www.gevent.org/">Gevent</a> is a well-known Python library for using green threads. Gevent is a green thread and non-blocking I/O. <code>gevent.monkey</code> modifies the behavior of standard Python libraries so that they allow non-blocking I/O operations to be performed.</p>
<p>Other libraries:</p>
<p>Let's see how performance changes if we start using green threads using gevent library in Python:</p>
<pre>
<code>import gevent.monkey
 
 # patch any other imported module that has a blocking code in it 
 # to make it asynchronous.
 gevent.monkey.patch_all()
 
 @timeit()
 def green_threaded(n_threads, func, *args):
 jobs = []
 for i in range(n_threads):
 jobs.append(gevent.spawn(func, *args))
 # ensure all jobs have finished execution
 gevent.wait(jobs)
 
 if __name__ == '__main__:
 ...
 green_threaded(10, cpu_bound, a, b)
 green_threaded(10, io_bound, urls)
 </code>
</pre>
<p>Results are: <code>cpu_bound</code> &mdash; 2.23 sec, <code>io_bound</code> &mdash; 6.85 sec.</p>
<p>Slower for CPU-bound function, and faster for I/O-bound function. As expected.</p>
<h3>Asyncio</h3>
<p>The <code>asyncio</code> package is described in the Python documentation as <a href="https://docs.python.org/3/library/asyncio.html">a library for writing parallel code</a>. However, <code>asyncio</code> is not multithreaded and is not multiprocessing. It is not built on top of one of them.</p>
<p>While Gevent and Twisted aim to be higher level frameworks, asyncio aims to be a lower-level implementation of an asynchronous event loop, with the intention that higher level frameworks like Twisted, Gevent or Tornado, will build on top of it. However, by itself, it makes a suitable framework on its own.</p>
<p>In fact, <code>asyncio</code> is a single-threaded, single-process project: it uses <a href="https://luminousmen.com/post/asynchronous-programming-cooperative-multitasking"><strong>cooperative multitasking</strong></a>. <code>asyncio</code> allows us to write asynchronous concurrent programs running in the same thread, using an event loop for scheduling tasks and multiplexing I/O through sockets (and other resources).</p>
<p><code>asyncio</code> provides us an event loop along with other good stuff. The event loop tracks different I/O events and switches to tasks which are ready and pauses the ones which are waiting on I/O. Thus we don&rsquo;t waste time on tasks which are not ready to run right now.</p>
<h3>How it works</h3>
<p>Synchronous and asynchronous functions/callables are different types &mdash; <strong>you can't just mix them</strong>. If you block a coroutine synchronously &mdash; maybe you use <code>time.sleep(10)</code> rather than <code>await asyncio.sleep(10)</code> &mdash; you don't return control to the event loop &mdash; the entire process will block.</p>
<p>You should think of your codebase as comprised of pieces of either sync code or async code &mdash; anything inside an <code>async def</code> is async code, anything else (including the main body of a Python file or class) is synchronous code.</p>
<p><img alt="asynchronous api" src="https://luminousmen.com/media/asynchronous-programming-await-the-future-4.jpg"></p>
<p>The idea is very simple. There&rsquo;s an event loop. And we have an asynchronous function (coroutine) in Python, you declare it with <code>async def</code>, which changes how its call behaves. In particular, <strong>calling it will immediately return a coroutine object</strong>, which basically says "I can run the coroutine and return a result when you await me".</p>
<p>We give those functions to the event loop and ask it to run them for us. The event loop gives us back a <code>Future</code> object, it&rsquo;s like <strong>a promise that we will get something back in the future</strong>. We hold on to the promise, time to time check if it has a value (when we feel impatient) and finally when the future has a value, we use it in some other operations.</p>
<p>When you call <code>await</code>, the function gets suspended while whatever you asked to wait on happens, and then when it's finished, the event loop will wake the function up again and resume it from the await call, passing any result out. Example:</p>
<pre>
<code>import asyncio
 
 async def say(what, when):
 await asyncio.sleep(when)
 print(what)
 
 loop = asyncio.get_event_loop()
 loop.run_until_complete(say('hello world', 1))
 loop.close()
 </code>
</pre>
<p>In the example here, the <code>say()</code> function pauses and gives back control to the event loop, which sees that <code>sleep</code> needs to run, calls it, and then that calls await and gets suspended with a marker to resume it one second. Once it resumes, <code>say</code> completes, returns a result, and then that makes main ready to run again and the event loop resumes it with the returned value.</p>
<p>This is how async code can have so many things happening at once - anything that's blocking calls await, and gets put onto the event loop's list of paused coroutines so something else can run. Everything that's paused has an associated callback that will wake it up again &mdash; some are time-based, some are I/O-based, and most of them are like the example above and waiting for a result from another coroutine.</p>
<p>Let's return to our example. We have two blocking functions <code>cpu_bound</code> and <code>io_bound</code>. As I said, we cannot mix synchronous and asynchronous operations &mdash; <strong>we must make all of them asynchronous</strong>. Naturally, not for everything there are asynchronous libraries. <strong>Some code remains <a href="https://luminousmen.com/post/asynchronous-programming-blocking-and-non-blocking">blocking</a></strong>, and it must somehow be run so that it does not block our event loop. For this, there is a good <code>run_in_executor()</code> method, which runs what we passed to it in one of the threads of the built-in pool, without blocking the main thread with the event loop. We will use this functionality for our CPU-bound function. We will rewrite the I/O-bound function completely to await those moments when we are waiting for an event.</p>
<pre>
<code>import asyncio
 import aiohttp
 
 async def async_func(N, func, *args):
 coros = [func(*args) for _ in range(N)]
 # run awaitable objects concurrently
 await asyncio.gather(*coros)
 
 
 async def a_cpu_bound(a, b):
 result = await loop.run_in_executor(None, cpu_bound, a, b)
 return result
 
 
 async def a_io_bound(urls):
 # create a coroutine function where we will download from individual url
 async def download_coroutine(session, url):
 async with session.get(url, timeout=10) as response:
 await response.text()
 
 # set an aiohttp session and download all our urls
 async with aiohttp.ClientSession(loop=loop) as session:
 for url in urls:
 await download_coroutine(session, url)
 
 
 if __name__ == '__main__':
 ...
 loop = asyncio.get_event_loop()
 with timeit():
 loop.run_until_complete(async_func(10, a_cpu_bound, a, b))
 
 with timeit():
 loop.run_until_complete(async_func(10, a_io_bound, urls))
 </code>
</pre>
<p>Results are: <code>cpu_bound</code> &mdash; 2.23 sec, <code>io_bound</code> &mdash; 4.37 sec.</p>
<p>Slower for CPU-bound function, and almost twice as fast as the threaded example for I/O-bound function.</p>
<h2>Making the Right Choice</h2>
<ul>
<li>CPU-bound -&gt; multiprocessing</li>
<li>I/O-bound, fast I/O, Limited Number of Connections -&gt; multithreading</li>
<li>I/O-bound, slow I/O, many connections -&gt; asyncio</li>
</ul>
<h2>Conclusion</h2>
<p>Threads will be easier if you have a typical web application that does not depend on external services, and relatively finite amount users for whom the response time will be predictably short.</p>
<p>async is suitable if the application spends most of the time reading/writing data, rather than processing it. For example, you have a lot of slow requests &mdash; websockets, long polling, or there are slow external synchronous backends, requests for which are unknown when they end.</p>
<p>Synchronous programming is what most often begins the development of applications in which sequential execution of commands is performed.</p>
<p>Even with conditional branching, loops, and function calls, we think about code in terms of performing one step at a time. After the completion of the current step, it proceeds to the next.</p>
<p>An asynchronous application behaves differently. It still running one step at a time, but the difference is that the system moving forward, it's not waiting for the completion of the current execution step. As a result, we are going to <a href="https://en.wikipedia.org/wiki/Event-driven_programming">the event-driven programming</a>.</p>
<p>asyncio is a great library and it's cool that it was included into Python standard library. asyncio has already begun to build an ecosystem (aiohttp, asyncpg, etc.) for application development. There are other event loop implementations (<a href="https://github.com/MagicStack/uvloop">uvloop</a>, <a href="https://github.com/dabeaz/curio">dabeaz/curio</a>, <a href="https://github.com/python-trio/trio">python-trio/trio</a>), and I think the asyncio will evolve in even more powerful tool.</p>
<h3>Links</h3>
<a href="https://www.buymeacoffee.com/OheL3Mfnp" target="blank">Support author</a>
</div>
</div>
<br>
<br>
<br>
<br>
<br>
<br>
via Blog | iamluminousmen<br>
<br>
June 28, 2019 at 11:49AM
